Support Gentoo mirrors in apt-cacher-ng

Please note the use of a script to build the gentoo_mirrors file live; this
should probably be done when each release is being built.

The geo-balanced http://distfiles.gentoo.org/ mirror is NOT included in the
mirrors3.xml source, so it gets added manually.

The first 3 sed expresssions aren't actually required, but are useful in making
the output list understandable.

Signed-off-by: Robin H. Johnson <robbat2@gentoo.org>

diff -Nuar --exclude '*.gz' apt-cacher-ng-0.7.19.orig/conf/acng.conf apt-cacher-ng-0.7.19/conf/acng.conf
--- apt-cacher-ng-0.7.19.orig/conf/acng.conf	2013-07-28 11:09:43.000000000 -0700
+++ apt-cacher-ng-0.7.19/conf/acng.conf	2013-12-03 19:55:07.000000000 -0800
@@ -44,6 +44,7 @@
 Remap-fedora:  file:fedora_mirrors # Fedora Linux
 Remap-epel:   file:epel_mirrors # Fedora EPEL
 Remap-slrep:  file:sl_mirrors # Scientific Linux
+Remap-gentoo: file:gentoo_mirrors.gz http://distfiles.gentoo.org/ /gentoo ; file:backends_gentoo # Gentoo Archives
 
 # This is usually not needed for security.debian.org because it's always the
 # same DNS hostname. However, it might be enabled in order to use hooks,
diff -Nuar --exclude '*.gz' apt-cacher-ng-0.7.19.orig/conf/backends_gentoo.default apt-cacher-ng-0.7.19/conf/backends_gentoo.default
--- apt-cacher-ng-0.7.19.orig/conf/backends_gentoo.default	1969-12-31 16:00:00.000000000 -0800
+++ apt-cacher-ng-0.7.19/conf/backends_gentoo.default	2013-12-03 19:46:38.000000000 -0800
@@ -0,0 +1,4 @@
+# This is a configuration file. All lines starting with # are ignored.
+
+# gentoo mirrors use geodns to dispatch to the nearest location
+http://distfiles.gentoo.org/
diff -Nuar --exclude '*.gz' apt-cacher-ng-0.7.19.orig/conf/gentoo_mirrors.sh apt-cacher-ng-0.7.19/conf/gentoo_mirrors.sh
--- apt-cacher-ng-0.7.19.orig/conf/gentoo_mirrors.sh	1969-12-31 16:00:00.000000000 -0800
+++ apt-cacher-ng-0.7.19/conf/gentoo_mirrors.sh	2013-12-04 18:20:07.000000000 -0800
@@ -0,0 +1,11 @@
+#!/bin/sh
+# This fetchs the live Gentoo mirrors list
+# robbat2@gentoo.org - 2013/Dec/03
+OUTFILE=/etc/apt-cacher-ng/gentoo_mirrors
+wget --save-headers -q http://www.gentoo.org/main/en/mirrors3.xml -O - \
+	|sed -n \
+	-e '/^[A-Z]/{s,^,#,g;p}' \
+	-e '/<mirrorgroup/{s,^,\n#,g;p}' \
+	-e '/<name/{s,^,#,g;p}' \
+	-e '/<uri/{/protocol="http"/{s/.*<uri[^>]\+>//g;s/<\/uri>//g;p}}' \
+	| gzip -9 >${OUTFILE}.gz
